Add optional `MetricsCollector` to the client (#388)
profunktor/fs2-rabbit: site/docs/examples/client-metrics.md

---
layout: docs
title:  "Client Metrics"
number: 18
---

# Client Metrics

RabbitMQ Java Client supports metrics collection via [Dropwizard](https://www.rabbitmq.com/blog/2016/11/30/metrics-support-in-rabbitmq-java-client-4-0/) or [Micrometer](https://www.rabbitmq.com/blog/2018/04/10/rabbitmq-java-client-metrics-with-micrometer-and-datadog/).

> At the moment of writing both providers are in the `amqp-client` 5.9.0. Just instantiate one as it is shown below.

Reviewer: ```suggestion
At the moment of writing both providers are in the `amqp-client` 5.9.0. You can instantiate one as shown below.
```

⛏️ Avoid using the word "just" in documentation, as it's always either useless, condescending, or both.

Author: Great suggestion, thanks for pointing it out 👍

```scala
val registry            = new MetricRegistry

> val dropwizardCollector = new StandardMetricsCollector(registry)

Reviewer: 🤔 Do these need construction in `IO`? In `Resource`? I'd be surprised if they were pure

Author: Haha, I didn't find it that bad.

### `MetricRegistry`

https://github.com/dropwizard/metrics/blob/release/4.1.x/metrics-core/src/main/java/com/codahale/metrics/MetricRegistry.java#L60-L66

### `StandardMetricsCollector`

https://github.com/rabbitmq/rabbitmq-java-client/blob/master/src/main/java/com/rabbitmq/client/impl/StandardMetricsCollector.java#L50-L70

Have you experienced any weirdness around this initialisers?

Reviewer 2: It is usually a good practice to suspend Java stuff because you never know :|

Reviewer: 👀 

Those classes contain mutable state; hence construction is not referentially transparent -- you can't factor `new` in/out freely without changing the meaning. It doesn't look like they have a lifecycle though, so no need for `Resource`

Author: Yeah, they are not resources.

I have no objection to wrap them in IO, I'll do it 👍 

By the way, they are out of IO in the http4s docs too: https://http4s.org/v0.18/middleware/ (scroll to the Dropwizard Metrics Middleware).

ease to use. 

```scala
object JmxReporterResource {
  def make[F[_]: Sync](registry: MetricRegistry): Resource[F, JmxReporter] = {
    val acquire = Sync[F].delay {
      val reporter = JmxReporter.forRegistry(registry).inDomain("com.rabbitmq.client.jmx").build
      reporter.start()
      reporter
    }

    val close = (reporter: JmxReporter) => Sync[F].delay(reporter.close()).void

    Resource.make(acquire)(close)
  }
}
```

Let's initialise the FS2 RabbitMQ client and AMQP channel with metrics.

```scala
val resources = for {
  blocker <- Blocker[IO]
  _       <- JmxReporterResource.make[IO](registry)
  client  <- Resource.liftF(RabbitClient[IO](config, blocker, metricsCollector = Some(dropwizardCollector)))
  channel <- client.createConnection.flatMap(client.createChannel)
} yield (channel, client)

val program = resources.use {
  case (channel, client) =>
    // Let's publish and consume and see the counters go up
}
```

The app is going to have now the following metrics under `com.rabbitmq.client.jmx`:
* Acknowledged messages
* Channels count
* Connections count
* Consumed messages
* Published messages
* Rejected messages

> ## Full Listing

Reviewer: 🤔 It might be worth us making an examples/ project to hold these kinds of things, but doesn't have to be in this PR

Author: An example Main file was added in the [examples](https://github.com/profunktor/fs2-rabbit/blob/cb3bb15ecb97d8945b14294310d84489d7926b0b/examples/src/main/scala/dev/profunktor/fs2rabbit/examples/DropwizardMetricsDemo.scala). Based on my experience, a lot of people scan the website to build their mental model of what it is first. Do you think it worth having it in the both places?

Reviewer: It would be nice if we could embed the file in the doc page, but I don't know if the tool we're using can do that.

It's fine as is I think

Author: Ah, yeah, I see what you mean. I couldn't find anything like that in the mdoc documentation too.
